game/scripts/vscripts/items/bloodmage.lua

```diff
@@ -0,0 +1,142 @@
+  -- remove it
+  local mod = caster:FindModifierByName( "modifier_item_bloodmage_sangromancy" )
+
+  if mod and not mod:IsNull() then
+    mod:Destroy()
+
+    -- caster:EmitSound( "OAA_Item.SiegeMode.Deactivate" )
+  else
+    -- if it isn't toggled, add the modifier and keep track of it
+    caster:AddNewModifier( caster, self, "modifier_item_bloodmage_sangromancy", {} )
+
+    -- caster:EmitSound( "OAA_Item.SiegeMode.Activate" )
+  end
+end
+
+--------------------------------------------------------------------------------
+
+modifier_item_bloodmage_sangromancy = class(ModifierBaseClass)
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:IsHidden()
+  return false
+end
+
+function modifier_item_bloodmage_sangromancy:IsDebuff()
+  return false
+end
+
+function modifier_item_bloodmage_sangromancy:IsPurgable()
+  return false
+end
+
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:OnCreated( event )
+  local spell = self:GetAbility()
+
+  spell.mod = self
+
+  self.spellamp = spell:GetSpecialValueFor( "sangromancy_spell_amp" )
+  self.selfDamage = spell:GetSpecialValueFor( "sangromancy_self_damage" )
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:OnRefresh( event )
+  local spell = self:GetAbility()
+
+  spell.mod = self
+
+  self.spellamp = spell:GetSpecialValueFor( "sangromancy_spell_amp" )
+  self.selfDamage = spell:GetSpecialValueFor( "sangromancy_self_damage" )
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:OnRemoved()
+  local spell = self:GetAbility()
+
+  if spell and not spell:IsNull() then
+    spell.mod = nil
+  end
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:DeclareFunctions()
+  local funcs = {
+    MODIFIER_PROPERTY_SPELL_AMPLIFY_PERCENTAGE,
+    MODIFIER_EVENT_ON_TAKEDAMAGE,
+  }
+
+  return funcs
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:GetModifierSpellAmplify_Percentage( event )
+  local spell = self:GetAbility()
+
+  return self.spellamp or spell:GetSpecialValueFor( "sangromancy_spell_amp" )
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_item_bloodmage_sangromancy:OnTakeDamage(event)
+  if event.damage_category == 0 and event.attacker == self:GetCaster() and not (event.unit == self:GetCaster()) then
+
+    local damage = {
+      victim = event.attacker,
+      attacker = event.attacker,
+      damage = event.original_damage * (self.selfDamage / 100),
+      damage_type = event.damage_type,
+      ability = event.inflictor,
```

Comment:
Sourcing the damage to the item (`self:GetAbility()`) might make more sense. This sources the damage to the spell that was cast (kinda like Double Edge).

Reply:
Won't this remove stuff like piercing spell immunity on some spells?

Reply:
That's not dependent on damage source. Calling `ApplyDamage` without any checks like this would actually just apply the damage indiscriminately. All magic damage would still be blocked by magic immunity either way.

Reply:
Never mind it doesn't work this anyways. 
And stuff like a piercing magic missile on blademail also isn't reflected to vengeful spirit if she has bkb activated